### PR TITLE
janet: improve stability of JPM and modules

### PR DESCRIPTION
**Pull request by pyrmont**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

The current formula for Janet installs the Janet Project Manager ("JPM"). JPM is Janet's official mechanism for installing Janet modules. Janet modules are bundles of code that can be either libaries or executables. Libraries can be source code in Janet that is loaded by a Janet project directly or source code in other languages that is compiled to use Janet's C interface. Similarly, executables can be source code in Janet that is run as executable scripts or source code in other languages that is compiled to an executable binary on installation.

Homebrew's formula installs JPM in the keg where Janet itself is installed. This means that JPM uses the `bin`, `lib` and `share/man` directories within the keg for modules that _it_ downloads. This is problematic for three reasons. First, it means that when Janet is updated, libraries installed previously seem to the user to have have disappeared. Second, executables are not on the user's PATH after installation unless the user add's the keg's `bin` directory to their path (which will in turn break if the respective keg is uninstalled). Third, man pages installed by JPM do not appear unless the user add's the keg's `share/man` directory to their manpath.

As with other formulae that install package managers (e.g. Node, Python), a solution that is more stable across runtime updates is for the package manager to use one or more directories outside the keg. This commit does that by configuring JPM during the bootstrap phase to use the respective directories within `{HOMEBREW_PREFIX}`. The downside of this approach is that when Janet is uninstalled completely, the files and directories created by JPM remain. Nevertheless, this approach is preferred to the status quo.

In addition to improving JPM stability, this commit also removes the build-time dependency on Meson and Ninja. Janet comes with a well-designed Makefile that works consistently across macOS and Linux without the need for other build tools.

### CODE REVIEW

**Comment by pyrmont** [comment]
@carlocab I'm not sure why the checksum changed. This is the checksum that homebrew on my local machine suggested. I didn't realise that would be a problem to update it. How would you like me to handle this?

**Review by chenrui333 — APPROVED** [review]
THanks!

File: Formula/j/janet.rb
Comment (chenrui333) [inline Formula/j/janet.rb:72]: why not just this?

```suggestion
    assert_match HOMEBREW_PREFIX/"lib/janet", shell_output("#{jpm} show-paths")
```

File: Formula/j/janet.rb
Comment (chenrui333) [inline Formula/j/janet.rb:33]: maybe just comment out?
```suggestion
    inreplace "Makefile", /^.*?\bmkdir\b.*?\$\(JANET_PATH\).*?$/, "#"
```

File: Formula/j/janet.rb
Comment (pyrmont) [reply Formula/j/janet.rb:71]: `assert_match` appears to need something that can respond to `#=~` which Pathname apparently doesn't. I can do either:

```ruby
assert_match (HOMEBREW_PREFIX/"lib/janet").to_s, shell_output("#{jpm} show-paths")
```

or

```ruby
assert_match "lib/janet", shell_output("#{jpm} show-paths")
```

The latter doesn't test that JPM has been put in the correct place but the former possibly is more resource intensive given it converts the Pathname into a String? (Although maybe that's not too bad if the Pathname has a String backing it—not sure.)

Do you have a preference?

Comment (carlocab) [reply Formula/j/janet.rb:71]: Probably just the former. But please do so in a separate PR, because this one already has bottles published.

File: Formula/j/janet.rb
Comment (carlocab) [reply Formula/j/janet.rb:32]: Not sure this one really matters, but if necessary it should be done in a follow-up.

**Comment by carlocab** [comment]
I'm resolving the comments to allow this to be merged, but please fix any remaining issues in a separate PR.

File: Formula/j/janet.rb
Comment (pyrmont) [reply Formula/j/janet.rb:32]: @chenrui333 Do you mean just match against `/^.*?\bmkdir\b.*?\S\(JANET_PATH\)/` and not bother with matching to the end of the line? Or is there a function I can use to insert a character into the first column of the line that has matched?

**Comment by carlocab** [comment]
Oops. Looks like I need to wait until this is properly merged to un-resolve those conversations.

File: Formula/j/janet.rb
Comment (chenrui333) [reply Formula/j/janet.rb:71]: 👍

File: Formula/j/janet.rb
Comment (chenrui333) [reply Formula/j/janet.rb:32]: yeah, I was just thinking about this comment `# Line removed by Homebrew formula` (it does not be homebrew specific though? 🤔 )

File: Formula/j/janet.rb
Comment (pyrmont) [reply Formula/j/janet.rb:71]: OK, thanks. Will make a separate PR.

File: Formula/j/janet.rb
Comment (pyrmont) [reply Formula/j/janet.rb:32]: Well, it's necessary to comment it out for Homebrew because the effect of this line in the Makefile is that it tries to create a directory outside of the keg (given the prefix I'm using) and the Homebrew sandbox won't allow that.

Comment (pyrmont) [reply Formula/j/janet.rb:32]: So it seemed helpful to note that this was specifically because of Homebrew's needs and not something broken in the Makefile itself.

**Pull request merged**
